### PR TITLE
New Feature: Language Selection

### PR DESCRIPTION
**Pull request by jlcases**
Greetings,

I've developed an enhanced feature for multi-language support:

Language Selection: This feature provides the capability to select from up to 7 different languages.
image

Language Selection Interface

Optimized Queries: I've implemented a total of 8 queries - 4 in English and 4 in the user-selected language to enhance search results.

Localized Reports: The system now generates reports in the user-selected language.
image

Report Interface

I believe this feature will greatly benefit the project by offering more versatility to users. I'm eager to contribute this to the main repository and am open to any suggestions or modifications you might have.

Thank you for considering my contribution.

Best regards,
Jose Luis Cases

### CODE REVIEW

**Review by ill-yes — CHANGES_REQUESTED** [review]
Looks good! But please write comment in code in english.

File: agent/prompts.py [hunk 1]
```diff
@@ -1,89 +1,84 @@
-def generate_agent_role_prompt(agent):
+def generate_agent_role_prompt(agent, language):
```
Comment (notvicent3) [inline agent/prompts.py:1]: touché

File: agent/prompts.py [hunk 2]
```diff
@@ -1,89 +1,84 @@
...
     """ Generates the agent role prompt.
-    Args: agent (str): The type of the agent.
-    Returns: str: The agent role prompt.
+    Args: 
+        agent (str): The type of the agent.
+        language (str): The language in which the answers should be provided.
```
Comment (notvicent3) [inline agent/prompts.py:5]: Touché.

File: agent/prompts.py [hunk 3]
```diff
@@ -1,89 +1,84 @@
...
-        "Computer Security Analyst Agent": "You are an AI specializing in computer security analysis. Your principal duty is to generate comprehensive, meticulously detailed, impartial, and systematically structured reports on computer security topics. This includes Exploits, Techniques, Threat Actors, and Advanced Persistent Threat (APT) Groups. All produced reports should adhere to the highest standards of scholarly work and provide in-depth insights into the complexities of computer security.",
-        "Default Agent": "You are an AI critical thinker research assistant. Your sole purpose is to write well written, critically acclaimed, objective and structured reports on given text."
+        "Finance Agent": f"You are a seasoned finance analyst AI assistant. Your primary goal is to compose comprehensive, astute, impartial, and methodically arranged financial reports based on provided data and trends. All answers must be in {language}.",
```
Comment (notvicent3) [inline agent/prompts.py:10]: Habría que adaptarlo (si es que ha cambiado a los prompts iniciales) pero Touché.

File: agent/prompts.py [hunk 4]
```diff
@@ -1,89 +1,84 @@
...
+        "Computer Security Analyst Agent": f"You are an AI specializing in computer security analysis. Your principal duty is to generate comprehensive, meticulously detailed, impartial, and systematically structured reports on computer security topics. This includes Exploits, Techniques, Threat Actors, and Advanced Persistent Threat (APT) Groups. All produced reports should adhere to the highest standards of scholarly work and provide in-depth insights into the complexities of computer security. All answers must be in {language}.",
+        "Default Agent": f"You are an AI critical thinker research assistant. Your sole purpose is to write well written, critically acclaimed, objective and structured reports on given text. All answers must be in {language}."
     }
 
-    return prompts.get(agent, "No such agent")
+    return prompts.get(agent, f"No such agent. All answers must be in {language}.")
```
Comment (notvicent3) [inline agent/prompts.py:18]: Touché.

File: agent/prompts.py [hunk 5]
```diff
@@ -1,89 +1,84 @@
...
 
 
-def generate_report_prompt(question, research_summary):
+
+def generate_report_prompt(question, research_summary,language):
```
Comment (notvicent3) [inline agent/prompts.py:22]: Touché. No sé realmente si la falta del espacio de research_summary, language): 
Cambiaría más allá de estética.

File: agent/prompts.py [hunk 6]
```diff
@@ -1,89 +1,84 @@
...
-    return f'"""{research_summary}""" Using the above information, answer the following'\
+    return f'"""{research_summary}""" Using the above information, answer  in "{language}" the following'\
            f' question or topic: "{question}" in a detailed report --'\
            " The report should focus on the answer to the question, should be well structured, informative," \
            " in depth, with facts and numbers if available, a minimum of 1,200 words and with markdown syntax and apa format. "\
             "You MUST determine your own concrete and valid opinion based on the given information. Do NOT deter to general and meaningless conclusions." \
            "Write all used source urls at the end of the report in apa format"
 
-def generate_search_queries_prompt(question):
+def generate_search_queries_prompt(question, language):
```
Comment (notvicent3) [inline agent/prompts.py:36]: Touché.

File: agent/prompts.py [hunk 7]
```diff
@@ -1,89 +1,84 @@
...
     """ Generates the search queries prompt for the given question.
     Args: question (str): The question to generate the search queries prompt for
     Returns: str: The search queries prompt for the given question
     """
 
-    return f'Write 4 google search queries to search online that form an objective opinion from the following: "{question}"'\
-           f'You must respond with a list of strings in the following format: ["query 1", "query 2", "query 3", "query 4"]'
-
+    return f'For the topic "{question}", list 4 search queries in English and 4 in {language}.'
```
Comment (notvicent3) [inline agent/prompts.py:42]: Touché.

File: agent/prompts.py [hunk 8]
```diff
@@ -1,89 +1,84 @@
...
 
-def generate_resource_report_prompt(question, research_summary):
-    """Generates the resource report prompt for the given question and research summary.
 
-    Args:
-        question (str): The question to generate the resource report prompt for.
-        research_summary (str): The research summary to generate the resource report prompt for.
+def generate_resource_report_prompt(question, research_summary, language):
```
Comment (notvicent3) [inline agent/prompts.py:45]: Touché.

File: agent/prompts.py [hunk 9]
```diff
@@ -1,89 +1,84 @@
...
-           f' question or topic: "{question}". The report should provide a detailed analysis of each recommended resource,' \
-           ' explaining how each source can contribute to finding answers to the research question.' \
-           ' Focus on the relevance, reliability, and significance of each source.' \
-           ' Ensure that the report is well-structured, informative, in-depth, and follows Markdown syntax.' \
-           ' Include relevant facts, figures, and numbers whenever available.' \
-           ' The report should have a minimum length of 1,200 words.'
 
-
-def generate_outline_report_prompt(question, research_summary):
+def generate_outline_report_prompt(question, research_summary,language):
```
Comment (notvicent3) [inline agent/prompts.py:56]: Touché.

File: agent/prompts.py [hunk 10]
```diff
@@ -1,89 +1,84 @@
...
-
-    return f'"""{research_summary}""" Using the above information, generate an outline for a research report in Markdown syntax'\
-           f' for the following question or topic: "{question}". The outline should provide a well-structured framework'\
-           ' for the research report, including the main sections, subsections, and key points to be covered.' \
-           ' The research report should be detailed, informative, in-depth, and a minimum of 1,200 words.' \
-           ' Use appropriate Markdown syntax to format the outline and ensure readability.'
-
-def generate_concepts_prompt(question, research_summary):
+    return (f'"""{research_summary}""" Using the above information, generate  '
+            'an outline in {language} for a research report in Markdown syntax'
```
Comment (notvicent3) [inline agent/prompts.py:63]: Touché.

File: agent/prompts.py [hunk 11]
```diff
@@ -1,89 +1,84 @@
...
     """ Generates the concepts prompt for the given question.
     Args: question (str): The question to generate the concepts prompt for
             research_summary (str): The research summary to generate the concepts prompt for
     Returns: str: The concepts prompt for the given question
     """
-
-    return f'"""{research_summary}""" Using the above information, generate a list of 5 main concepts to learn for a research report'\
-           f' on the following question or topic: "{question}". The outline should provide a well-structured framework'\
-           'You must respond with a list of strings in the following format: ["concepts 1", "concepts 2", "concepts 3", "concepts 4, concepts 5"]'
+    return (f'"""{research_summary}""" Using the above information, generate in {language}, '
```
Comment (notvicent3) [inline agent/prompts.py:75]: Touché.

File: agent/prompts.py [hunk 12]
```diff
@@ -1,89 +1,84 @@
...
+            ' a list of 5 main concepts to learn for a research report'
+            f' on the following question or topic: "{question}". The outline should provide a well-structured framework'
+            'You must respond with a list of strings in the following format: ["concepts 1", "concepts 2", "concepts 3", "concepts 4, concepts 5"]')
 
 
-def generate_lesson_prompt(concept):
+def generate_lesson_prompt(concept,language):
```
Comment (notvicent3) [inline agent/prompts.py:81]: Touché.

File: agent/prompts.py [hunk 13]
```diff
@@ -92,7 +87,7 @@ def generate_lesson_prompt(concept):
         str: The lesson prompt for the given concept.
     """
 
-    prompt = f'generate a comprehensive lesson about {concept} in Markdown syntax. This should include the definition'\
+    prompt = f'generate  a comprehensive lesson in {language} about {concept} in Markdown syntax. This should include the definition'\
```
Comment (notvicent3) [inline agent/prompts.py:90]: Touché.

File: agent/research_agent.py [hunk 14]
```diff
@@ -23,15 +23,16 @@
...
+    def __init__(self, question, agent, agent_role_prompt, language, websocket):
         """ Initializes the research assistant with the given question.
         Args: question (str): The question to research
         Returns: None
         """
 
         self.question = question
         self.agent = agent
-        self.agent_role_prompt = agent_role_prompt if agent_role_prompt else prompts.generate_agent_role_prompt(agent)
+        self.language = language  # New field for language
```
Comment (notvicent3) [inline agent/research_agent.py:34]: Touché.

File: agent/research_agent.py [hunk 15]
```diff
@@ -23,15 +23,16 @@
...
+        self.agent_role_prompt = agent_role_prompt if agent_role_prompt else prompts.generate_agent_role_prompt(agent, self.language)
```
Comment (notvicent3) [inline agent/research_agent.py:35]: Touché.

File: agent/research_agent.py [hunk 16]
```diff
@@ -90,10 +92,23 @@ async def create_search_queries(self):
         Args: None
         Returns: list[str]: The search queries for the given question
         """
-        result = await self.call_agent(prompts.generate_search_queries_prompt(self.question))
-        print(result)
-        await self.websocket.send_json({"type": "logs", "output": f"🧠 I will conduct my research based on the following queries: {result}..."})
-        return json.loads(result)
+        result = await self.call_agent(prompts.generate_search_queries_prompt(self.question,self.language))
```
Comment (notvicent3) [inline agent/research_agent.py:95]: Touché.

File: agent/research_agent.py [hunk 17]
```diff
@@ -90,10 +92,23 @@ async def create_search_queries(self):
...
+        queries = []
+
+        for line in lines:
+        # Busca las comillas en cada línea para extraer la consulta
+            start = line.find('"')
+            end = line.rfind('"')
+            if start != -1 and end != -1:
+                queries.append(line[start+1:end])
+
+        await self.websocket.send_json({"type": "logs", "output": f"🧠 I will conduct my research based on the following queries: {queries}..."})
```
Comment (notvicent3) [inline agent/research_agent.py:108]: Deberíamos pensar en introducir la interacción en el language seleccionado también. Touché.

File: agent/research_agent.py [hunk 18]
```diff
@@ -167,8 +183,8 @@ async def write_report(self, report_type, websocket):
         report_type_func = prompts.get_report_by_type(report_type)
         await websocket.send_json(
             {"type": "logs", "output": f"✍️ Writing {report_type} for research task: {self.question}..."})
-        answer = await self.call_agent(report_type_func(self.question, self.research_summary), stream=True,
-                                       websocket=websocket)
+        answer = await self.call_agent(report_type_func(self.question, self.research_summary, self.language), stream=True, websocket=websocket)
```
Comment (notvicent3) [inline agent/research_agent.py:186]: Touché.

File: agent/run.py [hunk 19]
```diff
@@ -34,21 +34,22 @@ async def disconnect(self, websocket: WebSocket):
         del self.sender_tasks[websocket]
         del self.message_queues[websocket]
 
-    async def start_streaming(self, task, report_type, agent, agent_role_prompt, websocket):
-        report, path = await run_agent(task, report_type, agent, agent_role_prompt, websocket)
+    async def start_streaming(self, task, report_type, agent, agent_role_prompt, language, websocket):
+        report, path = await run_agent(task, report_type, agent, agent_role_prompt,language,websocket)
```
Comment (notvicent3) [inline agent/run.py:38]: Touché.

File: agent/run.py [hunk 20]
```diff
@@ -34,21 +34,22 @@ async def disconnect(self, websocket: WebSocket):
         del self.sender_tasks[websocket]
         del self.message_queues[websocket]
 
-    async def start_streaming(self, task, report_type, agent, agent_role_prompt, websocket):
-        report, path = await run_agent(task, report_type, agent, agent_role_prompt, websocket)
+    async def start_streaming(self, task, report_type, agent, agent_role_prompt, language, websocket):
```
Comment (notvicent3) [inline agent/run.py:37]: Touché.

File: agent/run.py [hunk 21]
```diff
@@ -34,21 +34,22 @@ async def disconnect(self, websocket: WebSocket):
...
+        report, path = await run_agent(task, report_type, agent, agent_role_prompt,language,websocket)
         return report, path
 
 
-async def run_agent(task, report_type, agent, agent_role_prompt, websocket):
+async def run_agent(task, report_type, agent, agent_role_prompt, language, websocket):  # New field for language
```
Comment (notvicent3) [inline agent/run.py:42]: Touché.

File: agent/run.py [hunk 22]
```diff
@@ -34,21 +34,22 @@ async def disconnect(self, websocket: WebSocket):
...
     check_openai_api_key()
 
     start_time = datetime.datetime.now()
 
     # await websocket.send_json({"type": "logs", "output": f"Start time: {str(start_time)}\n\n"})
 
-    assistant = ResearchAgent(task, agent, agent_role_prompt, websocket)
+    assistant = ResearchAgent(task, agent, agent_role_prompt, language, websocket)  # Pasar el idioma al constructor del agente
```
Comment (notvicent3) [inline agent/run.py:49]: Touché.

File: client/index.html [hunk 23]
```diff
@@ -81,6 +81,22 @@ <h1 class="text-4xl font-extrabold mx-auto lg:text-7xl">
                 <option value="outline_report">Outline Report</option>
             </select>
         </div>
+        <div class="form-group">
+            <label for="language" class="agent-question">Select the output language for the report:</label>
```
Comment (notvicent3) [inline client/index.html:85]: Touché.

File: client/scripts.js [hunk 24]
```diff
@@ -29,11 +29,13 @@ const GPTResearcher = (() => {
         const task = document.querySelector('input[name="task"]').value;
         const report_type = document.querySelector('select[name="report_type"]').value;
         const agent = document.querySelector('input[name="agent"]:checked').value;
-  
+        const language = document.getElementById("language").value;  // get language from dropdown
```
Comment (notvicent3) [inline client/scripts.js:32]: Touché.

File: client/scripts.js [hunk 25]
```diff
@@ -29,11 +29,13 @@ const GPTResearcher = (() => {
...
+        
         const requestData = {
           task: task,
           report_type: report_type,
           agent: agent,
+          language: language  // add language to request data
```
Comment (notvicent3) [inline client/scripts.js:38]: Touché.

File: main.py [hunk 26]
```diff
@@ -13,6 +13,8 @@ class ResearchRequest(BaseModel):
     task: str
     report_type: str
     agent: str
+    language: str  # New fiedl for language
```
Comment (notvicent3) [inline main.py:16]: Touché.

File: main.py [hunk 27]
```diff
@@ -47,6 +49,7 @@ async def websocket_endpoint(websocket: WebSocket):
                 task = json_data.get("task")
                 report_type = json_data.get("report_type")
                 agent = json_data.get("agent")
+                language = json_data.get("language")  # New field for language
```
Comment (notvicent3) [inline main.py:52]: Touché.

File: main.py [hunk 28]
```diff
@@ -57,7 +60,7 @@ async def websocket_endpoint(websocket: WebSocket):
 
                 await websocket.send_json({"type": "logs", "output": f"Initiated an Agent: {agent}"})
                 if task and report_type and agent:
-                    await manager.start_streaming(task, report_type, agent, agent_role_prompt, websocket)
+                    await manager.start_streaming(task, report_type, agent, agent_role_prompt, language, websocket) # New field for language
```
Comment (notvicent3) [inline main.py:63]: Touché.

**Review by notvicent3 — COMMENTED** [review]
Touché.

**Comment by assafelovic** [comment]
Languages are by default when using most LLMs so selecting the lang is redundant.

**Comment by assafelovic** [comment]
@saccohuo definitely in demand but very old and many conflicts. Should we open a new PR for this? What ever you decide and happy to help and push this to production! @jlcases

**Comment by Gallisatyricon** [comment]
I would love to get this function. English is always coming out of the queries, even when the input is a long phrase in another language. Tks in advance if this becomes possible.

**Comment by tisDDM** [comment]
Is anybody working on this topic? 
If it is stale and it is still wanted I will give it a go, because I also need results in two non-English languages.

GPTR unfortunately is reverting many times into writing English reports, which makes it very uncomfortable for international use.

**Comment by assafelovic** [comment]
@tisDDM @Gallisatyricon happy to update I've just pushed a fix for languages. Simply export the prefered language as an env var and it should work. For example `export LANGUAGE=spanish`.
Make sure to pull latest!

**Comment by assafelovic** [comment]
See latest PR here: https://github.com/assafelovic/gpt-researcher/pull/1026

**Comment by assafelovic** [comment]
Closed, added basic support here: https://github.com/assafelovic/gpt-researcher/pull/1026

**Pull request closed without merging**